filament/include/filament/RenderableManager.h

```diff
@@ -340,6 +342,50 @@ class UTILS_PUBLIC RenderableManager : public FilamentAPI {
         Builder& skinning(size_t boneCount, Bone const* bones) noexcept; //!< \overload
         Builder& skinning(size_t boneCount) noexcept; //!< \overload
 
+        /**
+         * Define bone indices and weights for vertex skinning. The first pair value defines index
+         * of the bone and the second value is the bone weight. The pairs substitute \c BONE_INDICES
+         * and the \c BONE_WEIGHTS defined in the VertexBuffer. Both ways of indices and weights
+         * definition must not be combined in one primitive.
+         * Number of pairs is not limited to 4 bones per vertex.
+         * All bone weights of one vertex should sum to one. Otherwise they will be normalized.
+         * Data must be rectangular and number of bone pairs must be same for all vertices of this
+         * primitive.
+         * The data is arranged sequentially, all bone pairs for the first vertex, then for the
+         * second vertex, and so on.
+         *
+         * @param primitiveIndex zero-based index of the primitive, must be less than the primitive
+         *                       count passed to Builder constructor
+         * @param indicesAndWeights pairs of bone index and bone weight for all vertices
+         *                          sequentially
+         * @param count number of all pairs, must be a multiple of vertexCount of the primitive
+         * @param bonesPerVertex number of bone pairs, same for all vertices of the primitive
+         *
+         * @return Builder reference for chaining calls.
+         */
+        Builder& boneIndicesAndWeights(size_t primitiveIndex,math::float2 const* indicesAndWeights, size_t count, size_t bonesPerVertex) noexcept;
```

Comment:
Thanks for your note, I've been thinking about it and discussing it with colleagues. Moving the boneIndicesAndWeights method to VertexBuffer::Builder has several disadvantages.
The VertexBuffer::Builder method, especially attribute(), only defines the structure of the vertex buffer, not the data itself. The boneIndicesAndWeights method defines the data and vertices attributes are defined according to it. In my opinion, this violates the philosophy of the api.
Even if we moved this method to VertexBuffer::Builder, some additional data processing would have to be done in RenderebleManager::Builder, because that's where the texture with the bone indices and weights is baked in. The texture is common to all geometries, it must be created after all bone data is defined for all vertex buffers.
How exactly do you plan to design that API?

Reply:
This problem escaped me during the first read; `RenderableManager::Builder` is actually modifying/setting data on `VertexBuffer`, right? 

Could we have a whole new `BoneBuffer`, where this API is moved to `BoneBuffer::Builder` and then a `BoneBuffer` can be set on a `VertexBuffer` (not on the Builder) and also set on `RenderableManabger::Builder`?

It would make the API look like this:

```
auto bones = BoneBuffer::Builder().boneIndicesAndWeights(...).build();
auto renderable = RenderableManager::Builder()...geometry(vertexBuffer, ...).boneBuffer(bones).build();
vertexBuffer.setBoneBuffer(bones);
```

Would that work?

Reply:
Yes, RenderableManager:Builder is modifyingVertexBuffer. 
This is because when the VertexBuffer is built, we don't know if the vertices have bone indices and weights attached. Vertex buffer is built without VertexAttribute::BONE_INDICES and VertexAttribute::BONE_WEIGHTS attributes.
After the indices and weights are set in RenderableManager::Builder().boneIndicesAndWeights(...).build() method, the vertex buffer must be rebuilt.

We need something that would determine that the vertex buffer uses the bone vertex attributes for skinning. But using the standard attribute() method for this is strange.

There should be something that set the vertex attributes VertexAttribute::BONE_INDICES and VertexAttribute::BONE_WEIGHTS without attaching a buffer. Then there will be no need to build the vertex buffer again. And there is no need for a special Bone buffer either.


auto vertexBuffer = VertexBuffer::Builder()
…
.attribute(VertexAttribute::BONE_INDICES, ???)
.attribute(VertexAttribute::BONE_WEIGHTS, ???)
.build();

auto renderable = RenderableManager::Builder()
…
.geometry(vertexBuffer, ...)
.skinning(...)
.boneIndicesAndWeights(...)
.build();

When building a vertex buffer, it is necessary to know that attributes for skinning are needed. Then it is no longer necessary to rebuild the vertex buffer.
How should the API method for setting the attribute look like?

VertexBuffer::Builder().setBoneAttribute()

Reply:
I'm okay with something like `VertexBuffer::Builder().setBoneAttribute()` but what does that mean for the existing API? meaning, we need to make sure we're not breaking the existing API.
Maybe instead of `setBoneAttribute()` it could be `VertexBuffer::Builder().advancedSkinning()`?

Reply:
That sounds good, I'll redo it that way, thanks for your opinion